### PR TITLE
Example of integration test using pytest

### PR DESCRIPTION
**Pull request by jameshadfield**
John here's a proof-of-concept way to test the CLI using pytest to consider (I'm sure there are plenty of others). Can run via `py.test tests/cli/test_tree.py` (you'll need to do `pip install more-itertools` first)

### CODE REVIEW

File: tests/cli/test_tree.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+# and https://docs.pytest.org/en/latest/reference.html#testdir
+
+pytest_plugins = "pytester"
+
+@pytest.fixture
+def run(testdir):
+    def do_run(*args):
+        args = ["augur", "tree"] + list(args)
+        return testdir.run(*args)
+    return do_run
```
Comment (Zsailer) [inline tests/cli/test_tree.py:16]: This is really interesting. I haven't seen `pytester` used in this way. Clever! Usually pytester is used to test pytest plugins, but this is pretty cool. 

Have you seen [pytest-console-scripts](https://github.com/kvas-it/pytest-console-scripts)? I've used it in a few places around the Jupyter stack. Though, I'm not sure how it works with output *files* that you want to read and assert its contents.

Comment (jameshadfield) [reply tests/cli/test_tree.py:16]: This was _literally_ the first thing I've done in pytest, so it was a steep learning curve!

Comment (Zsailer) [reply tests/cli/test_tree.py:16]: pytest is super clean but can be difficult to follow. Fixtures are often black boxes!

**Comment by elebow** [comment]
I think this is the right approach in general, but that expressing expectations in the test code is unnecessary extra work that harms maintainability.

Since Augur has a CLI and straightforward, deterministic behavior, the integration test suite can simply be a set of input data and expected output files. Call it declarative testing, perhaps.

Something like this, in shell script:

```sh
#!/bin/sh

augur filter --sequences tests/integration/input_files/filter_vcf -o temp.vcf
if diff temp.vcf tests/integration/expected_outputs/filter_out.vcf; then
	exit 1
fi

augur mask --whatever -o temp.vcf
if diff temp.vcf tests/integration/expected_outputs/mask_out.vcf; then
	exit 1
fi

...
```

Or in pytest, to take advantage of parallelization and nicer output:

```python
    def test_filter(self):
        os.system("augur filter --sequences tests/integration/input_files/filter.vcf -o temp.vcf")
        assert filecmp.cmp("temp.vcf", "tests/integration/expected_outputs/filter.vcf")

    def test_mask(self):
        os.system("augur mask --whatever -o temp.vcf")
        assert filecmp.cmp("temp.vcf", "tests/integration/expected_outputs/mask.vcf")

...
```

The test pipelines already do something very close to this. We would just want to verify each intermediate file against a known good copy.

**Comment by huddlej** [comment]
After a lot of discussion and trying different approaches, we've decided to use the Cram test framework for the integration/function/end-to-end tests. Check out [the documentation for how to write and run these tests](https://github.com/nextstrain/augur/blob/master/DEV_DOCS.md#testing) for more information.

**Pull request closed without merging**